include/deal.II/lac/la_vector.templates.h

```diff
@@ -303,51 +396,33 @@ namespace LinearAlgebra
                                      const VectorSpaceVector<Number> &V,
                                      const VectorSpaceVector<Number> &W)
   {
-    this->add(a,V);
-
-    return *this*W;
-  }
-
-
-
-  template <typename Number>
-  typename VectorSpaceVector<Number>::real_type Vector<Number>::l1_norm_recursive(
-    unsigned int i,
-    unsigned int j) const
-  {
-    Assert(j>=i, ExcInternalError());
-    typename ReadWriteVector<Number>::real_type norm = 0.;
-
-    if ((j-i)!=0)
-      {
-        norm += l1_norm_recursive(i,(i+j)/2);
-        norm += l1_norm_recursive((i+j)/2+1,j);
-      }
-    else
-      norm += std::abs(this->val[i]);
-
-    return norm;
-  }
-
-
+    // Check that casting will work.
+    Assert(dynamic_cast<const Vector<Number>*>(&V)!=NULL,
+           ExcVectorTypeNotCompatible());
+    // Check that casting will work.
+    Assert(dynamic_cast<const Vector<Number>*>(&W)!=NULL,
+           ExcVectorTypeNotCompatible());
 
-  template <typename Number>
-  typename VectorSpaceVector<Number>::real_type Vector<Number>::l2_norm_squared_recursive(
-    unsigned int i,
-    unsigned int j) const
-  {
-    Assert(j>=i, ExcInternalError());
+    // Downcast V. If fails, throws an exception.
+    const Vector<Number> &down_V = dynamic_cast<const Vector<Number>&>(V);
+    // Downcast W. If fails, throws an exception.
+    const Vector<Number> &down_W = dynamic_cast<const Vector<Number>&>(W);
+    AssertIsFinite(a);
+    Assert(down_V.size()==this->size(),
+           ExcMessage("Cannot add two vectors with different numbers of elements"));
+    Assert(down_W.size()==this->size(),
+           ExcMessage("Cannot add two vectors with different numbers of elements"));
 
-    typename ReadWriteVector<Number>::real_type norm = 0.;
-    if ((j-i)!=0)
-      {
-        norm += l2_norm_squared_recursive(i,(i+j)/2);
-        norm += l2_norm_squared_recursive((i+j)/2+1,j);
-      }
-    else
-      norm += std::pow(std::abs(this->val[i]),2);
+    Number sum;
+    internal::AddAndDot<Number> adder;
+    adder.X = this->val;
+    adder.a = a;
+    adder.V = down_V.val;
+    adder.W = down_W.val;
+    internal::parallel_reduce(adder, this->size(), sum, this->thread_loop_partitioner);
```

Comment:
Of course, this would apply to all functors that we use in the `vector_operations_internal.h` file.